pagegraph/src/graph_algos.rs

```diff
@@ -431,6 +432,50 @@ impl PageGraph {
         }
     }
 
+    /// Get a collection of all Resource nodes whose requests match a set of adblock filter patterns.
+    pub fn resources_matching_filters(&self, patterns: Vec<String>) -> Vec<(NodeId, &Node)> {
+        let source_url = self.root_url();
+
+        let source_url = url::Url::parse(&source_url).expect("Could not parse source URL");
+        let source_hostname = source_url.host_str().expect(&format!("Source URL has no host, {:?}", source_url));
+        let source_domain = get_domain(source_hostname);
+        let blocker = Engine::from_rules(&patterns);
+
+        self.nodes
+            .iter()
+            .filter(|(id, node)| match &node.node_type {
+                NodeType::Resource { url } => {
+                    let request_url = match url::Url::parse(url) {
+                        Ok(request_url) => request_url,
+                        Err(_) => return false,
+                    };
+                    let request_url_hostname = match request_url.host_str() {
+                        Some(host) => host,
+                        None => return false,
+                    };
+                    let request_url_domain = get_domain(request_url_hostname);
+
+                    let request_types = self.resource_request_types(id);
+
+                    request_types.iter().map(|(request_type, _size)|  {
+                        let third_party = if source_domain.is_empty() {
+                            None
+                        } else {
+                            Some(source_domain != request_url_domain)
+                        };
+                        blocker
+                            .check_network_urls_with_hostnames(url, request_url_hostname,
+                                                               source_hostname,
+                                                               request_type,
+                                                               third_party).matched
+                    }).any(|matched| matched)
+                }
+                _ => false
+            })
+            .map(|(id, node)| (id.clone(), node))
+            .collect()
+    }
+
     /// Get a collection of all Resource nodes whose requests match a given adblock filter pattern.
     pub fn resources_matching_filter(&self, pattern: &str) -> Vec<(NodeId, &Node)> {
```

Comment:
Last request, I promise :sweat_smile:

Can you make `resources_matching_filter` just call `resources_matching_filters(vec![pattern.to_string()])`? Then we can deduplicate a lot of code. I realize it's not 100% the same implementation, checking against the single filter versus compiling it into an engine, but it shouldn't matter in practice.

Reply:
> Last request, I promise 😅

No thank you for this excellent review! Helps avoid bugs. I pushed that change.